### ISSUE TITLE
Log and MaskedArray

### ISSUE THREAD

**Issue by brossignol** [numpy/numpy#12183]
Masked array do not work with numpy.log, and numpy.log10

>> A = numpy.ma.MaskedArray([1, 0, 2], mask =[False, True, False])
>> numpy.log(A)

-> RuntimeWarning: divide by zero encountered in log

**Comment by eric-wieser** [comment]
Works for me:
```
>>>np.log(A)
RuntimeWarning: divide by zero encountered in log
masked_array(data = [0.0 -- 0.6931471805599453],
             mask = [False  True False],
       fill_value = 1e+20)
```

**Comment by mattip** [comment]
maybe you would be better served by `where`:
```
a = np.array([1, 0, 2])
out = np.full(a.shape, np.nan)
np.log(a, where=(a != 0), out=out)
```

**Comment by brossignol** [comment]
@eric-wieser 
np.log return the right result, but it give a useless warning.
We do not have this warning with >>1 / A, for exemple.

**Comment by eric-wieser** [comment]
@brossignol: That's a fair comment - try and include a more complete description of the problem in the issue next time :)

**Comment by eric-wieser** [comment]
> We do not have this warning with `1 / A`

Note that the warning _does_ appear with:

```
>>> np.divide(1, A)
RuntimeWarning: divide by zero encountered in true_divide
masked_array(data = [1.0 -- 0.5],
             mask = [False  True False],
       fill_value = 1e+20)
```

**Comment by WarrenWeckesser** [comment]
I agree with @brossignol: `np.log(np.ma.masked_array([1, 0, 2], mask=[False, True, False]))` should not generate a warning.  Neither should `np.divide(1.0, np.ma.masked_array([1, 0, 2], mask=[False, True, False]))`.  Note that `1.0 / np.ma.masked_array([1, 0, 2], mask=[False, True, False])` does not generate a warning.

Masked values should be completely ignored; if the underlying implementation actually operates on *all* the values in the array (perhaps for efficiency), and thus sets floating point exception flags when it computes `log(0)` or `1/0`, the code should know that those flags can be ignored and cleared when they are set because of an operation on a masked value.

**Comment by eric-wieser** [comment]
> the code should know that those flags can be ignored 

Only if those flags are generated by the masked values. If they're generated by a non-masked value, then they ought to be preserved (or converted to a mask).

The problem is, that comes with a performance hit, as the `where=` argument to ufuncs is slow.

**Comment by WarrenWeckesser** [comment]
@eric-wieser wrote:
> That comes with a performance hit...

Can this be done in a way that only the masked arrays take a performance hit?  The warnings are a real nuisance that make it difficult to treat masked arrays as first classes citizens.

**Comment by melissawm** [comment]
Note that this causes the example from the docs [here](https://numpy.org/devdocs/reference/maskedarray.generic.html#operations-on-masked-arrays) to have an undocumented warning:

```python
>>> from numpy import ma
>>> x = ma.array([-1, 1, 0, 2, 3], mask=[0, 0, 0, 0, 1])
>>> import numpy as np
>>> np.log(x)
__main__:1: RuntimeWarning: divide by zero encountered in log
__main__:1: RuntimeWarning: invalid value encountered in log
masked_array(data=[--, 0.0, --, 0.6931471805599453, --],
             mask=[ True, False,  True, False,  True],
       fill_value=1e+20)
>>> 
```

**Comment by fengluoqiuwu** [comment]
It seems that this warning still occurs in the latest version. I will look into it and try to fix it.

**Comment by greatofdream** [comment]
> It seems that this warning still occurs in the latest version.

I think the invalid of mask array for `numpy.log10` should be indicated in the documentation.
